Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +90,769 @@ local macro "pderiv_simp" : tactic =>
+    div_sub_div _ _ (pow_ne_zero 2 hPz) (pow_ne_zero 2 hQz), mul_comm <| _ ^ 2, addZ]
+  field_simp [mul_ne_zero (mul_ne_zero hPz hQz) <| sub_ne_zero_of_ne hx,
+    mul_ne_zero (mul_ne_zero (pow_ne_zero 3 hPz) (pow_ne_zero 3 hQz)) <| sub_ne_zero_of_ne hx]
+  ring1
+
+variable (V) in
+/-- The $X$-coordinate of the addition of two distinct point representatives. -/
+def addX (P Q : Fin 3 → R) : R :=
+  P x * Q x ^ 2 * P z ^ 2 - 2 * P y * Q y * P z * Q z + P x ^ 2 * Q x * Q z ^ 2
+    - V.a₁ * P x * Q y * P z ^ 2 * Q z - V.a₁ * P y * Q x * P z * Q z ^ 2
+    + 2 * V.a₂ * P x * Q x * P z ^ 2 * Q z ^ 2 - V.a₃ * Q y * P z ^ 4 * Q z
+    - V.a₃ * P y * P z * Q z ^ 4 + V.a₄ * Q x * P z ^ 4 * Q z ^ 2 + V.a₄ * P x * P z ^ 2 * Q z ^ 4
+    + 2 * V.a₆ * P z ^ 4 * Q z ^ 4
+
+lemma addX_eq' {P Q : Fin 3 → R} (hP : V.Equation P) (hQ : V.Equation Q) :
+    V.addX P Q * (P z * Q z) ^ 2 =
+      (P y * Q z ^ 3 - Q y * P z ^ 3) ^ 2
+        + V.a₁ * (P y * Q z ^ 3 - Q y * P z ^ 3) * P z * Q z * addZ P Q
+        - V.a₂ * P z ^ 2 * Q z ^ 2 * addZ P Q ^ 2 - P x * Q z ^ 2 * addZ P Q ^ 2
+        - Q x * P z ^ 2 * addZ P Q ^ 2 := by
+  linear_combination (norm := (rw [addX, addZ]; ring1)) -Q z ^ 6 * (equation_iff P).mp hP
+    - P z ^ 6 * (equation_iff Q).mp hQ
+
+lemma addX_eq {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) : W.addX P Q =
+      ((P y * Q z ^ 3 - Q y * P z ^ 3) ^ 2
+        + W.a₁ * (P y * Q z ^ 3 - Q y * P z ^ 3) * P z * Q z * addZ P Q
+        - W.a₂ * P z ^ 2 * Q z ^ 2 * addZ P Q ^ 2 - P x * Q z ^ 2 * addZ P Q ^ 2
+        - Q x * P z ^ 2 * addZ P Q ^ 2) / (P z * Q z) ^ 2 := by
+  rw [← addX_eq' hP hQ, mul_div_cancel_right₀ _ <| pow_ne_zero 2 <| mul_ne_zero hPz hQz]
+
+lemma addX_smul (P Q : Fin 3 → R) (u v : Rˣ) :
+    V.addX (u • P) (v • Q) = ((u * v) ^ 2) ^ 2 * V.addX P Q := by
+  simp only [addX, smul_fin3_ext]
+  ring1
+
+lemma addX_of_Z_eq_zero_left {P Q : Fin 3 → R} (hPz : P z = 0) :
+    V.addX P Q = (P x * Q z) ^ 2 * Q x := by
+  rw [addX, hPz]
+  ring1
+
+lemma addX_of_Z_eq_zero_right {P Q : Fin 3 → R} (hQz : Q z = 0) :
+    V.addX P Q = (-(Q x * P z)) ^ 2 * P x := by
+  rw [addX, hQz]
+  ring1
+
+lemma addX_of_X_eq' {P Q : Fin 3 → R} (hP : V.Equation P) (hQ : V.Equation Q)
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) :
+    V.addX P Q * (P z * Q z) ^ 2 = (P y * Q z ^ 3 - Q y * P z ^ 3) ^ 2 := by
+  simp only [addX_eq' hP hQ, addZ_of_X_eq hx, add_zero, sub_zero, mul_zero, zero_pow two_ne_zero]
+
+lemma addX_of_X_eq {P Q : Fin 3 → F} (hP : W.Equation P) (hQ : W.Equation Q) (hPz : P z ≠ 0)
+    (hQz : Q z ≠ 0) (hx : P x * Q z ^ 2 = Q x * P z ^ 2) :
+    W.addX P Q = (-((P y * Q z ^ 3 - Q y * P z ^ 3) / (P z * Q z))) ^ 2 := by
+  rw [neg_sq, div_pow, ← addX_of_X_eq' hP hQ hx,
+    mul_div_cancel_right₀ _ <| pow_ne_zero 2 <| mul_ne_zero hPz hQz]
+
+private lemma toAffine_addX_of_ne {P Q : Fin 3 → F} {n d : F} (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
```

Comment:
Again what is the logic of all these `private` lemmas?

Reply:
These are auxiliary lemmas for `toAffine_add/dblX/Y_of_eq`, which are essentially the same lemma but with an actual sane expression on the right hand side that is useful.